utils/resolved_udp_conn_test.go

```diff
@@ -0,0 +1,298 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package utils
+
+import (
+	"fmt"
+	"net"
+	"runtime"
+	"syscall"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+	"github.com/uber/jaeger-client-go/log"
+)
+
+type mockResolver struct {
+	mock.Mock
+}
+
+func (m *mockResolver) ResolveUDPAddr(network string, hostPort string) (*net.UDPAddr, error) {
+	args := m.Called(network, hostPort)
+
+	a0 := args.Get(0)
+	if a0 == nil {
+		return (*net.UDPAddr)(nil), args.Error(1)
+	}
+	return a0.(*net.UDPAddr), args.Error(1)
+}
+
+type mockDialer struct {
+	mock.Mock
+}
+
+func (m *mockDialer) DialUDP(network string, laddr, raddr *net.UDPAddr) (*net.UDPConn, error) {
+	args := m.Called(network, laddr, raddr)
+
+	a0 := args.Get(0)
+	if a0 == nil {
+		return (*net.UDPConn)(nil), args.Error(1)
+	}
+
+	return a0.(*net.UDPConn), args.Error(1)
+}
+
+func newUDPConn() (net.PacketConn, *net.UDPConn, error) {
+	mockServer, err := net.ListenPacket("udp", "127.0.0.1:0")
+	if err != nil {
+		return nil, nil, err
+	}
+
+	addr, err := net.ResolveUDPAddr("udp", mockServer.LocalAddr().String())
+	if err != nil {
+		mockServer.Close()
+		return nil, nil, err
+	}
+
+	conn, err := net.DialUDP("udp", nil, addr)
+	if err != nil {
+		mockServer.Close()
+		return nil, nil, err
+	}
+
+	return mockServer, conn, nil
+}
+
+func TestNewResolvedUDPConn(t *testing.T) {
+	hostPort := "blahblah:34322"
+
+	mockServer, clientConn, err := newUDPConn()
+	if !assert.NoError(t, err) {
+		t.FailNow()
+	}
+	defer mockServer.Close()
+
+	resolver := mockResolver{}
+	resolver.
+		On("ResolveUDPAddr", "udp", hostPort).
+		Return(&net.UDPAddr{
+			IP:   net.IPv4(123, 123, 123, 123),
+			Port: 34322,
+		}, nil).
+		Once()
+
+	dialer := mockDialer{}
+	dialer.
+		On("DialUDP", "udp", (*net.UDPAddr)(nil), &net.UDPAddr{
+			IP:   net.IPv4(123, 123, 123, 123),
+			Port: 34322,
+		}).
+		Return(clientConn, nil).
+		Once()
+
+	conn, err := newResolvedUDPConn(hostPort, time.Hour, resolver.ResolveUDPAddr, dialer.DialUDP, log.NullLogger)
+	assert.NoError(t, err)
+	assert.NotNil(t, conn)
+
+	err = conn.Close()
+	assert.NoError(t, err)
+
+	// assert the actual connection was closed
+	assert.Error(t, clientConn.Close())
+
+	resolver.AssertExpectations(t)
+	dialer.AssertExpectations(t)
+}
+
+func TestResolvedUDPConnWrites(t *testing.T) {
+	hostPort := "blahblah:34322"
+
+	mockServer, clientConn, err := newUDPConn()
+	if !assert.NoError(t, err) {
+		t.FailNow()
+	}
+	defer mockServer.Close()
+
+	resolver := mockResolver{}
+	resolver.
+		On("ResolveUDPAddr", "udp", hostPort).
+		Return(&net.UDPAddr{
+			IP:   net.IPv4(123, 123, 123, 123),
+			Port: 34322,
+		}, nil).
+		Once()
+
+	dialer := mockDialer{}
+	dialer.
+		On("DialUDP", "udp", (*net.UDPAddr)(nil), &net.UDPAddr{
+			IP:   net.IPv4(123, 123, 123, 123),
+			Port: 34322,
+		}).
+		Return(clientConn, nil).
+		Once()
```

Comment:
nit: these seem repeated from previous test, could they be combined?
```
resolved, dialer := mockResolverAndDialer(hostPort, clientConn)
```

Reply:
i think it may be hard to combine these into a function because each test has a different set of calls being mocked